crypto/bio/bss_conn.c

```diff
@@ -188,8 +188,12 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
         case BIO_CONN_S_CONNECT:
             BIO_clear_retry_flags(b);
             ERR_set_mark();
-            ret = BIO_connect(b->num, BIO_ADDRINFO_address(c->addr_iter),
-                              BIO_SOCK_KEEPALIVE | c->connect_mode);
+
+            opts = c->connect_mode;
+            if (BIO_ADDRINFO_socktype(c->addr_iter) != SOCK_DGRAM)
+                opts |= BIO_SOCK_KEEPALIVE;
```

Comment:
Its a corner case, but  I don't believe this will work for SOCK_SEQ_PACKET, which is the socket type for sctp.  suggest that it might be better for keepalive to be default off with opt-in (i.e. iff BIO_ADDRINFO_socktype() == SOCK_STREAM)